components/centraldashboard/app/api.ts

```diff
@@ -181,33 +268,170 @@ export class Api {
-            return;
-          }
-
-          const profile = req.body as CreateProfileRequest;
-          try {
-            // Use the request body if provided, fallback to auth headers
-            await this.profilesService.createProfile({
-              metadata: {
-                name: profile.namespace || req.user.username,
-              },
-              spec: {
-                owner: {
-                  kind: 'User',
-                  name: profile.user || req.user.email,
+              if (!req.user.hasAuth) {
+                return apiError({
+                  res,
+                  code: 405,
+                  error: ERRORS.operation_not_supported,
+                });
+              }
+    
+              const profile = req.body as CreateProfileRequest;
+              try {
+                const namespace = profile.namespace || req.user.username;
+                // Use the request body if provided, fallback to auth headers
+                await this.profilesService.createProfile({
+                  metadata: {
+                    name: namespace,
+                  },
+                  spec: {
+                    owner: {
+                      kind: 'User',
+                      name: profile.user || req.user.email,
+                    }
+                  },
+                });
+                res.json({message: `Created namespace ${namespace}`});
+              } catch (err) {
+                this.surfaceProfileControllerErrors({
+                  res,
+                  msg: 'Unexpected error creating profile',
+                  err,
+                });
+              }
+            })
+        .post(
+              '/add-contributor/:namespace',
+              async (req: Request, res: Response) => {
+                const {namespace} = req.params;
+                const {contributor} = req.body as AddOrRemoveContributorRequest;
+                if (!contributor || !namespace) {
+                  return apiError({
+                    res,
+                    error: `Missing contributor / namespace fields.`,
+                  });
+                }
+                if (!EMAIL_RGX.test(contributor)) {
+                  return apiError({
+                    res,
+                    error: `Contributor doesn't look like a valid email address`,
+                  });
+                }
+                let errIndex = 0;
+                try {
+                  const binding = this.mapSimpleBindingToWorkgroupBinding({
+                    user: contributor,
+                    namespace,
+                    role: 'contributor',
+                  });
+                  const {headers} = req;
+                  delete headers['content-length'];
+                  await this.profilesService
+                    .createBinding(binding, {headers});
+                  errIndex++;
+                  const users = await this.getContributors(namespace);
```

Comment:
Is it necessary to reload the contributors? If the action succeeds, the UI would be responsible for adding/removing the item from view right?

Reply:
This is great because instead of the UI updating just the one contributor, it also syncs the rest of the changes in case other changes were made. Every operation to your contributor list makes sure you're in sync with the absolute truth. It also leads to a simpler handler logic as the response for all the APIs are directly property bound to the view.